## Step 1: Loading the Dataset

In this step, we load the SACCO audited accounts dataset into our environment.

The dataset is stored in the `data/` folder within our project directory.

Loading the data allows us to begin exploring its structure and preparing it for analysis.

We start with a single Excel file before scaling to multiple datasets.

In [1]:
import pandas as pd

df = pd.read_excel("data/Acumen Audited Accounts 2023.xlsx")

## Step 2: Previewing the Dataset

After loading the dataset, we inspect the first few rows to understand its structure.

This helps us:
- Get a quick view of the data
- Understand how columns are organized
- Identify any obvious issues (missing values, formatting)

This is the first step in understanding the dataset before deeper analysis.

In [2]:
df.head()

,Unnamed: 0,ACUMEN SACCO SOCIETY,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,...,Unnamed: 24,Unnamed: 25,Unnamed: 26,Unnamed: 27,Unnamed: 28,Unnamed: 29,Unnamed: 30,Unnamed: 31,Unnamed: 32,Unnamed: 33
0,NaN,STATEMENT OF FINANCIAL POSITION,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,FINANCIAL YEAR 2021,Aud-2019,Aud-2020,Jan,Feb,Mar,Apr,May,Jun,...,2022-08-01 00:00:00,2022-09-01 00:00:00,2022-10-01 00:00:00,2022-11-01 00:00:00,2022-12-01 00:00:00,audited 2022,NaN,NaN,NaN,NaN
2,1,Cash & Cash Equivalent,10834463,60942575,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,52281693,58048024,62794651,63055663.28,77262741,77254741,NaN,NaN,NaN,NaN
3,1.1,Cash in hand(both local and foreign notes and ...,250946,1162644,0,0,0,0,0,0,...,12187787,17550823,21997216,15032040.68,1229050,1229050.75,NaN,NaN,NaN,NaN
4,1.2,Cash at bank:(placement with financial institu...,10583517,59779931,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,40093906,40497201,40797435,48023622.6,76033691,76025690.25,NaN,NaN,NaN,NaN


## Step 3: Initial Data Issues

From the preview, the dataset is not in a clean tabular format.

Key issues observed:
- Column names are labeled as "Unnamed"
- The first few rows contain titles and metadata instead of actual data
- Data appears to start from a lower row

This is common in financial reports exported from Excel.

To proceed, we need to:
- Identify the correct header row
- Reload the dataset using the correct structure

In [3]:
df = pd.read_excel("data/Acumen Audited Accounts 2023.xlsx", skiprows=2)
df.head()

,Unnamed: 0,FINANCIAL YEAR 2021,Aud-2019,Aud-2020,Jan,Feb,Mar,Apr,May,Jun,...,2022-08-01 00:00:00,2022-09-01 00:00:00,2022-10-01 00:00:00,2022-11-01 00:00:00,2022-12-01 00:00:00,audited 2022,Unnamed: 30,Unnamed: 31,Unnamed: 32,Unnamed: 33
0,1,Cash & Cash Equivalent,10834463.0,60942575.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,52281693.0,58048024.0,62794651.0,63055663.28,77262741.0,77254741.00,NaN,NaN,NaN,NaN
1,1.1,Cash in hand(both local and foreign notes and ...,250946.0,1162644.0,0.00,0.00,0.00,0.00,0.00,0.00,...,12187787.0,17550823.0,21997216.0,15032040.68,1229050.0,1229050.75,NaN,NaN,NaN,NaN
2,1.2,Cash at bank:(placement with financial institu...,10583517.0,59779931.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,40093906.0,40497201.0,40797435.0,48023622.60,76033691.0,76025690.25,NaN,NaN,NaN,NaN
3,2,Prepayments & Sundry Receivables,8620453.0,11664214.0,0.00,0.00,0.00,0.00,0.00,0.00,...,0.0,0.0,0.0,0.00,0.0,11282181.00,NaN,-8289401.16,NaN,NaN
4,3,Financial Investments,115855.0,115855.0,41747825.78,42026698.00,42341640.92,35634084.84,35905690.98,36173046.56,...,115855.0,115855.0,115855.0,115855.00,115855.0,115855.00,NaN,NaN,NaN,NaN


## Step 4: Understanding Data Structure

After cleaning the header rows, the dataset now shows financial line items across multiple time periods.

Key observations:
- Each row represents a financial category (e.g. Cash, Investments)
- Columns represent different time periods (years and months)
- The dataset is in a wide format

To prepare for analysis, we will:
- Rename key columns
- Reshape the dataset into a structured format suitable for analysis

In [4]:
df = df.rename(columns={
    'Unnamed: 0': 'Account_ID',
    'FINANCIAL YEAR 2021': 'Account_Name'
})

df.head()

,Account_ID,FINANCIAL YEAR 2021,Aud-2019,Aud-2020,Jan,Feb,Mar,Apr,May,Jun,...,2022-08-01 00:00:00,2022-09-01 00:00:00,2022-10-01 00:00:00,2022-11-01 00:00:00,2022-12-01 00:00:00,audited 2022,Unnamed: 30,Unnamed: 31,Unnamed: 32,Unnamed: 33
0,1,Cash & Cash Equivalent,10834463.0,60942575.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,52281693.0,58048024.0,62794651.0,63055663.28,77262741.0,77254741.00,NaN,NaN,NaN,NaN
1,1.1,Cash in hand(both local and foreign notes and ...,250946.0,1162644.0,0.00,0.00,0.00,0.00,0.00,0.00,...,12187787.0,17550823.0,21997216.0,15032040.68,1229050.0,1229050.75,NaN,NaN,NaN,NaN
2,1.2,Cash at bank:(placement with financial institu...,10583517.0,59779931.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,40093906.0,40497201.0,40797435.0,48023622.60,76033691.0,76025690.25,NaN,NaN,NaN,NaN
3,2,Prepayments & Sundry Receivables,8620453.0,11664214.0,0.00,0.00,0.00,0.00,0.00,0.00,...,0.0,0.0,0.0,0.00,0.0,11282181.00,NaN,-8289401.16,NaN,NaN
4,3,Financial Investments,115855.0,115855.0,41747825.78,42026698.00,42341640.92,35634084.84,35905690.98,36173046.56,...,115855.0,115855.0,115855.0,115855.00,115855.0,115855.00,NaN,NaN,NaN,NaN


## Step 5: Cleaning Column Names

Some column names contained hidden spaces, which prevented proper renaming.

We cleaned the column names by removing extra spaces and then renamed key columns:
- Account_ID
- Account_Name

This ensures consistency and avoids errors in later steps.

In [5]:
#renaming again
df = df.rename(columns={
    'Unnamed: 0': 'Account_ID',
    'FINANCIAL YEAR  2021': 'Account_Name'
})

In [6]:
#checking again
df.head()

,Account_ID,Account_Name,Aud-2019,Aud-2020,Jan,Feb,Mar,Apr,May,Jun,...,2022-08-01 00:00:00,2022-09-01 00:00:00,2022-10-01 00:00:00,2022-11-01 00:00:00,2022-12-01 00:00:00,audited 2022,Unnamed: 30,Unnamed: 31,Unnamed: 32,Unnamed: 33
0,1,Cash & Cash Equivalent,10834463.0,60942575.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,52281693.0,58048024.0,62794651.0,63055663.28,77262741.0,77254741.00,NaN,NaN,NaN,NaN
1,1.1,Cash in hand(both local and foreign notes and ...,250946.0,1162644.0,0.00,0.00,0.00,0.00,0.00,0.00,...,12187787.0,17550823.0,21997216.0,15032040.68,1229050.0,1229050.75,NaN,NaN,NaN,NaN
2,1.2,Cash at bank:(placement with financial institu...,10583517.0,59779931.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,40093906.0,40497201.0,40797435.0,48023622.60,76033691.0,76025690.25,NaN,NaN,NaN,NaN
3,2,Prepayments & Sundry Receivables,8620453.0,11664214.0,0.00,0.00,0.00,0.00,0.00,0.00,...,0.0,0.0,0.0,0.00,0.0,11282181.00,NaN,-8289401.16,NaN,NaN
4,3,Financial Investments,115855.0,115855.0,41747825.78,42026698.00,42341640.92,35634084.84,35905690.98,36173046.56,...,115855.0,115855.0,115855.0,115855.00,115855.0,115855.00,NaN,NaN,NaN,NaN


## Step 6: Removing Irrelevant Columns

The dataset contains several empty or irrelevant columns labeled as "Unnamed".

These columns do not provide useful information for analysis and may introduce errors.

We remove these columns to simplify the dataset and focus only on meaningful data.

In [7]:
df.columns.str.contains('Unnamed')

Index([False, False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,   nan,   nan,   nan,
         nan,   nan,   nan,   nan,   nan,   nan,   nan,   nan,   nan, False,
        True,  True,  True,  True],
      dtype='object')

In [8]:
df = df.loc[:, ~df.columns.astype(str).str.contains('Unnamed')]

df.head()

,Account_ID,Account_Name,Aud-2019,Aud-2020,Jan,Feb,Mar,Apr,May,Jun,...,2022-04-22 00:00:00,2022-05-22 00:00:00,2022-06-22 00:00:00,2022-07-22 00:00:00,2022-08-01 00:00:00,2022-09-01 00:00:00,2022-10-01 00:00:00,2022-11-01 00:00:00,2022-12-01 00:00:00,audited 2022
0,1,Cash & Cash Equivalent,10834463.0,60942575.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,44314258.0,47378805.93,49968245.55,52249064.44,52281693.0,58048024.0,62794651.0,63055663.28,77262741.0,77254741.00
1,1.1,Cash in hand(both local and foreign notes and ...,250946.0,1162644.0,0.00,0.00,0.00,0.00,0.00,0.00,...,5403363.0,8177749.93,10487055.59,8472651.02,12187787.0,17550823.0,21997216.0,15032040.68,1229050.0,1229050.75
2,1.2,Cash at bank:(placement with financial institu...,10583517.0,59779931.0,17588524.65,17139917.86,16670330.97,20330140.74,21293871.83,24310141.01,...,38910895.0,39201056.00,39481189.96,43776413.42,40093906.0,40497201.0,40797435.0,48023622.60,76033691.0,76025690.25
3,2,Prepayments & Sundry Receivables,8620453.0,11664214.0,0.00,0.00,0.00,0.00,0.00,0.00,...,0.0,0.00,0.00,0.00,0.0,0.0,0.0,0.00,0.0,11282181.00
4,3,Financial Investments,115855.0,115855.0,41747825.78,42026698.00,42341640.92,35634084.84,35905690.98,36173046.56,...,115855.0,115855.00,115855.00,115855.00,115855.0,115855.0,115855.0,115855.00,115855.0,115855.00


## Step 7: Reshaping the Dataset (Wide to Long Format)

The dataset is currently in a wide format where each time period (months and years) is represented as a separate column.

This structure is not suitable for analysis or modeling.

To make the data usable, we convert it into a long format where:
- Each row represents a single observation
- Time periods are stored in one column
- Values are stored in another column

This transformation is essential for time series analysis and predictive modeling.

In [9]:
df_long = df.melt(
    id_vars=['Account_ID', 'Account_Name'],
    var_name='Date',
    value_name='Value'
)

df_long.head()

,Account_ID,Account_Name,Date,Value
0,1,Cash & Cash Equivalent,Aud-2019,10834463.0
1,1.1,Cash in hand(both local and foreign notes and ...,Aud-2019,250946.0
2,1.2,Cash at bank:(placement with financial institu...,Aud-2019,10583517.0
3,2,Prepayments & Sundry Receivables,Aud-2019,8620453.0
4,3,Financial Investments,Aud-2019,115855.0


## Step 8: Structured Dataset Achieved

The dataset has now been transformed into a long format.

Each row represents:
- A financial category (Account_Name)
- A specific time period (Date)
- A corresponding value (Value)

This structure allows us to:
- Analyze trends over time
- Compare financial categories
- Prepare the data for feature selection and modeling

In [10]:
df_long['Date'].unique()

array(['Aud-2019', 'Aud-2020', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
       'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec', 'Aud-21',
       datetime.datetime(2022, 1, 22, 0, 0),
       datetime.datetime(2022, 2, 22, 0, 0),
       datetime.datetime(2022, 3, 22, 0, 0),
       datetime.datetime(2022, 4, 22, 0, 0),
       datetime.datetime(2022, 5, 22, 0, 0),
       datetime.datetime(2022, 6, 22, 0, 0),
       datetime.datetime(2022, 7, 22, 0, 0),
       datetime.datetime(2022, 8, 1, 0, 0),
       datetime.datetime(2022, 9, 1, 0, 0),
       datetime.datetime(2022, 10, 1, 0, 0),
       datetime.datetime(2022, 11, 1, 0, 0),
       datetime.datetime(2022, 12, 1, 0, 0), 'audited 2022'], dtype=object)

## Step 9: Cleaning the Date Column

The Date column contains inconsistent formats, including:
- Year-based values (e.g. Aud-2019, audited 2022)
- Month names (e.g. Jan, Feb)
- Missing values

To make the dataset usable, we standardize the Date column into a consistent format.

This step is essential for time-based analysis and modeling.

In [11]:
import numpy as np

def clean_date(x):
    if pd.isna(x):
        return np.nan
    x = str(x).strip().lower()
    
    if "2019" in x:
        return "2019"
    elif "2020" in x:
        return "2020"
    elif "21" in x:
        return "2021"
    elif "2022" in x:
        return "2022"
    else:
        return x.capitalize()  # Jan, Feb, etc.

df_long['Clean_Date'] = df_long['Date'].apply(clean_date)

df_long[['Date', 'Clean_Date']].drop_duplicates()

,Date,Clean_Date
0,Aud-2019,2019
49,Aud-2020,2020
98,Jan,Jan
147,Feb,Feb
196,Mar,Mar
245,Apr,Apr
294,May,May
343,Jun,Jun
392,Jul,Jul
441,Aug,Aug


## Step 10: Removing Invalid Dates

Some rows contain missing date values.

These rows do not contribute to time-based analysis and may introduce errors.

We remove them to ensure the dataset is clean and consistent.

In [12]:
df_long = df_long.dropna(subset=['Clean_Date'])

df_long['Clean_Date'].unique()

array(['2019', '2020', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul',
       'Aug', 'Sep', 'Oct', 'Nov', 'Dec', '2021', '2022'], dtype=object)

## Step 11: Creating Features from Financial Categories

Currently, financial categories (Account_Name) exist as rows.

To prepare for analysis and modeling, we convert these into columns.

This transformation creates a structured dataset where:
- Each row represents a time period
- Each column represents a financial feature

In [13]:
df_pivot = df_long.pivot_table(
    index='Clean_Date',
    columns='Account_Name',
    values='Value',
    aggfunc='sum'
)

df_pivot.head()

Account_Name,Accounts Payable & Other Liabilities,Accounts Receivables,Adjustment to Equity,Allowance for Loan Loss,Balances with other SACCO Societies,Capital Grants,Cash & Cash Equivalent,Cash at bank:(placement with financial institutions),Cash in hand(both local and foreign notes and coins),Current Year's Surplus,...,Share Capital (Members share capital per Bylaws),"Short Term Deposits (e.g Fixed deposits, special savings)",Statutory Reserve,Tax Payable,Tax Recoverable,Total Assets,Total Deposit liabilities,Total Equity,Total Liabilities,Total Liabilities and Equity
Clean_Date,,,,,,,,,,,,,,,,,,,,,
2019,14172704.00,0.0,0.0,1.460356e+06,0.0,0.0,1.083446e+07,1.058352e+07,2.509460e+05,710651.00,...,1.727634e+07,552646.0,5.780960e+06,3610.0,0.0,2.693712e+08,2.216134e+08,3.358527e+07,2.357861e+08,2.693714e+08
2020,413121.00,0.0,0.0,2.131924e+07,0.0,0.0,6.094258e+07,5.977993e+07,1.162644e+06,17320.00,...,1.897013e+07,708204.0,5.780960e+06,14602.0,0.0,2.934117e+08,2.577053e+08,3.528895e+07,2.581184e+08,2.934073e+08
2021,19393076.00,0.0,0.0,1.141865e+07,0.0,0.0,5.668734e+07,5.550940e+07,1.177937e+06,15839579.00,...,2.067421e+07,968674.0,9.745185e+06,1046054.0,0.0,3.694371e+08,2.963373e+08,5.370675e+07,3.157304e+08,3.694371e+08
2022,95090710.59,0.0,0.0,2.189497e+08,0.0,0.0,6.297820e+08,5.280153e+08,1.017668e+08,59563865.15,...,2.993520e+08,0.0,1.282062e+08,5073986.0,0.0,4.992839e+09,4.103885e+09,7.938627e+08,4.198976e+09,4.992839e+09
Apr,563322.84,0.0,0.0,1.737470e+07,0.0,0.0,2.033014e+07,2.033014e+07,0.000000e+00,5756177.71,...,1.944109e+07,0.0,5.785290e+06,13002.0,0.0,3.179216e+08,2.694771e+08,4.788110e+07,2.700405e+08,3.179216e+08


## Step 12: Final Structured Dataset

The dataset has been successfully transformed into a structured format suitable for analysis.

Key characteristics:
- Each row represents a time period
- Each column represents a financial variable (feature)
- Values represent financial amounts

This structure allows for:
- Time series analysis
- Feature selection
- Predictive modeling

This marks the transition from data preparation to analytical modeling.

## Step 13: Filtering Annual Data

The dataset contains both monthly and yearly data.

To simplify analysis and ensure consistency, we focus on annual data only.

This reduces noise and allows for clearer trend analysis and modeling.

In [14]:
df_pivot = df_pivot.loc[['2019', '2020', '2021', '2022']]

df_pivot

Account_Name,Accounts Payable & Other Liabilities,Accounts Receivables,Adjustment to Equity,Allowance for Loan Loss,Balances with other SACCO Societies,Capital Grants,Cash & Cash Equivalent,Cash at bank:(placement with financial institutions),Cash in hand(both local and foreign notes and coins),Current Year's Surplus,...,Share Capital (Members share capital per Bylaws),"Short Term Deposits (e.g Fixed deposits, special savings)",Statutory Reserve,Tax Payable,Tax Recoverable,Total Assets,Total Deposit liabilities,Total Equity,Total Liabilities,Total Liabilities and Equity
Clean_Date,,,,,,,,,,,,,,,,,,,,,
2019,14172704.00,0.0,0.0,1.460356e+06,0.0,0.0,1.083446e+07,1.058352e+07,2.509460e+05,710651.00,...,1.727634e+07,552646.0,5.780960e+06,3610.0,0.0,2.693712e+08,2.216134e+08,3.358527e+07,2.357861e+08,2.693714e+08
2020,413121.00,0.0,0.0,2.131924e+07,0.0,0.0,6.094258e+07,5.977993e+07,1.162644e+06,17320.00,...,1.897013e+07,708204.0,5.780960e+06,14602.0,0.0,2.934117e+08,2.577053e+08,3.528895e+07,2.581184e+08,2.934073e+08
2021,19393076.00,0.0,0.0,1.141865e+07,0.0,0.0,5.668734e+07,5.550940e+07,1.177937e+06,15839579.00,...,2.067421e+07,968674.0,9.745185e+06,1046054.0,0.0,3.694371e+08,2.963373e+08,5.370675e+07,3.157304e+08,3.694371e+08
2022,95090710.59,0.0,0.0,2.189497e+08,0.0,0.0,6.297820e+08,5.280153e+08,1.017668e+08,59563865.15,...,2.993520e+08,0.0,1.282062e+08,5073986.0,0.0,4.992839e+09,4.103885e+09,7.938627e+08,4.198976e+09,4.992839e+09


## Step 14: Feature Selection 

Feature selection is based on actual available financial variables in the dataset.

We select key variables that represent:
- Financial strength
- Lending activity
- Liquidity
- Capital structure

These features are important for predicting SACCO performance.

In [15]:
#Step 15-Creating model dataset
features = [
    'Total Assets',
    'Total Deposit liabilities',
    'Total Equity',
    'Total Liabilities',
    'Cash & Cash Equivalent',
    'Financial Investments',
    'Gross Loan Portfolio',
    'Net Loan Portfolio',
    'Non withdrawable deposits (BOSA member deposits)',
    'Share Capital (Members share capital per Bylaws)',
    'Statutory Reserve'
]

target = "Current Year's Surplus"

In [16]:
list(df_pivot.columns)

['Accounts Payable & Other Liabilities',
 'Accounts Receivables',
 'Adjustment to Equity',
 'Allowance for Loan Loss',
 'Balances with other SACCO Societies',
 'Capital Grants',
 'Cash & Cash Equivalent',
 'Cash at bank:(placement with financial institutions)',
 'Cash in hand(both local and foreign notes and coins)',
 "Current Year's Surplus",
 'Deferred Tax Assets',
 'Deferred Tax Liability',
 'Dividends Payable',
 'External Borrowings',
 'Financial Investments',
 'Government Securities - Treasury Bills/bonds,  ',
 'Gross Loan Portfolio',
 'Intangible Assets',
 'Investment Properties',
 'Investments in companies - shares/stocks ',
 'Net Loan Portfolio',
 'Non withdrawable deposits (BOSA member deposits)',
 'Other Assets',
 'Other Equity Accounts',
 'Other Liabilities',
 'Other Reserves (provide notes on these reserves)',
 'Other Securities - Commercial papers/Bonds',
 'Prepaid Lease rentals',
 'Prepayments & Sundry Receivables',
 "Prior Years' Retained Earnings",
 'Property & Equipmen

In [17]:
#Build Model Dataset.
df_model = df_pivot[features + [target]]

df_model

Account_Name,Total Assets,Total Deposit liabilities,Total Equity,Total Liabilities,Cash & Cash Equivalent,Financial Investments,Gross Loan Portfolio,Net Loan Portfolio,Non withdrawable deposits (BOSA member deposits),Share Capital (Members share capital per Bylaws),Statutory Reserve,Current Year's Surplus
Clean_Date,,,,,,,,,,,,
2019,2.693712e+08,2.216134e+08,3.358527e+07,2.357861e+08,1.083446e+07,1.158550e+05,2.475976e+08,2.461372e+08,2.203165e+08,1.727634e+07,5.780960e+06,710651.00
2020,2.934117e+08,2.577053e+08,3.528895e+07,2.581184e+08,6.094258e+07,1.158550e+05,2.382020e+08,2.168827e+08,2.547207e+08,1.897013e+07,5.780960e+06,17320.00
2021,3.694371e+08,2.963373e+08,5.370675e+07,3.157304e+08,5.668734e+07,1.158550e+05,3.122795e+08,3.008609e+08,2.874705e+08,2.067421e+07,9.745185e+06,15839579.00
2022,4.992839e+09,4.103885e+09,7.938627e+08,4.198976e+09,6.297820e+08,1.165917e+08,4.126524e+09,3.907574e+09,3.950072e+09,2.993520e+08,1.282062e+08,59563865.15


STARTING POINT


Load Cleaned Data

In this step, we load cleaned SACCO financial data.

The data has been standardized into a consistent format:
- Account_Name
- Financial values across years

This ensures consistency and simplifies further analysis.

In [18]:
import pandas as pd
import glob

files = glob.glob("data_clean/*.csv")

df_list = []

for file in files:
    df = pd.read_csv(file)
    df_list.append(df)

len(df_list)

52

Combine Cleaned Datasets

After loading all cleaned SACCO datasets, the next step is to combine them into a single dataset.

Each file represents financial data from a different SACCO but follows the same standardized structure:
- Account_Name
- Financial values across years (2019–2022)

By concatenating these datasets:
- We create a unified dataset for analysis
- Enable comparison across multiple SACCOs
- Prepare the data for transformation and modeling

This step ensures that all SACCO data is consolidated into one dataframe for further processing.

In [19]:
df_all = pd.concat(df_list)

df_all.head()

,Account_Name,2019,2020,2021,2022,2012,2013,2014,2015,2016,...,Unnamed: 16375,Unnamed: 16376,Unnamed: 16377,Unnamed: 16378,Unnamed: 16379,Unnamed: 16380,Unnamed: 16381,Unnamed: 16382,Unnamed: 16383,Unnamed: 12
0,Total Assets,269371213.0,2.934117e+08,369437115.0,403387182.7,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Total Deposit liabilities,221613427.0,2.577053e+08,296337292.0,320546208.0,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Total Equity,33585269.0,3.528895e+07,53706746.0,63405036.0,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Total Liabilities,235786131.0,2.581184e+08,315730368.0,339982145.0,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Cash & Cash Equivalent,10834463.0,6.094258e+07,56687339.0,77254741.0,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Handling Missing Values

Some financial values were recorded as zeros but represented missing data.

To address this:
- Data was cleaned to ensure proper numeric formatting
- Columns were converted to numeric format in Python
- Zero values were replaced with missing values (NaN)

This ensures accurate representation of financial data and prevents misleading analysis.

In [20]:
df_all.iloc[:, 1:] = df_all.iloc[:, 1:].apply(pd.to_numeric, errors='coerce')

C:\Users\morin\AppData\Local\Temp\ipykernel_616\2289449922.py:1: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  df_all.iloc[:, 1:] = df_all.iloc[:, 1:].apply(pd.to_numeric, errors='coerce')


In [21]:
df_all.iloc[:, 1:] = df_all.iloc[:, 1:].mask(df_all.iloc[:, 1:] == 0)

Reshape Data for Analysis

The combined dataset is currently in a wide format, where each year is represented as a separate column.

To prepare the data for analysis and modeling, we reshape it into a long format. This transformation:
- Converts year columns into a single variable
- Creates a consistent structure for time-based analysis
- Simplifies further feature engineering and modeling

The resulting dataset will contain:
- Account_Name
- Year
- Value

In [22]:
df_long = df_all.melt(
    id_vars=['Account_Name'],
    var_name='Year',
    value_name='Value'
)

df_long.head()

,Account_Name,Year,Value
0,Total Assets,2019,269371213.0
1,Total Deposit liabilities,2019,221613427.0
2,Total Equity,2019,33585269.0
3,Total Liabilities,2019,235786131.0
4,Cash & Cash Equivalent,2019,10834463.0


Create Modeling Dataset with SACCO and Year

The dataset includes multiple SACCOs with varying time coverage.

To preserve data integrity and avoid mixing financial data across different SACCOs:
- A SACCO identifier is included
- Both SACCO and Year are used as unique identifiers

The dataset is reshaped into a wide format where:
- Each row represents a SACCO at a specific year
- Each column represents a financial feature (e.g., Total Assets, Loans, Equity)
- Values correspond to financial amounts

This structure ensures accurate analysis and supports predictive modeling.

In [23]:
import os
import pandas as pd

df_list = []

for file in files:
    df = pd.read_csv(file)
    
    # clean columns immediately
    df.columns = df.columns.str.strip()
    
    # remove unnamed columns
    df = df.loc[:, ~df.columns.str.contains('Unnamed')]
    
    # extract SACCO name
    sacco_name = os.path.splitext(os.path.basename(file))[0]
    
    # assign SACCO
    df['SACCO'] = sacco_name
    
    df_list.append(df)

df_all = pd.concat(df_list, ignore_index=True)

In [24]:
df_all[['SACCO']].head()

,SACCO
0,ACUMEN SACCO
1,ACUMEN SACCO
2,ACUMEN SACCO
3,ACUMEN SACCO
4,ACUMEN SACCO


In [25]:
#CONFIRMING THE SACCO NAMES
df_all['SACCO'].unique()

array(['ACUMEN SACCO', 'DEFENCE SACCO', 'ECO-PILLAR SACCO', 'EDIS SACCO',
       'FORTITUDE SACCO', 'GOLDEN PILLAR SACCO', 'GOODFAITH SACCO',
       'GOODWAY SACCO', 'HAZINA SACCO', 'HOME BUSINESS SACCO', 'IG SACCO',
       'IMARIKA SACCO', 'JAMII SACCO', 'JUMUIKA SACCO', 'K-PILLAR SACCO',
       'KENYA ACHIEVAS SACCO', 'KENYA BANKERS SACCO', 'KIMISITU SACCO',
       'KOLENGE SACCO', 'KORU SACCO', 'KWETU SACCO', 'KWIKAS SACCO',
       'LAINISHA SACCO', 'MAFANIKIO SACCO', 'MAGADI SACCO',
       'MAGEREZA SACCO', 'MENTOR SACCO', 'NYATI SACCO', 'PATNAS SACCO',
       'PRIMETIME SACCO', 'SIMBA CHAI SACCO', 'SIRAJI SACCO',
       'SMART LIFE SACCO', 'SOTICO SACCO', 'STIMA SACCO',
       'STRATEGIC DT SACCO', 'TABASAMU SACCO', 'TEMBO SACCO',
       'TOWER RETURNS', 'TRANS COUNTIES SACCO',
       'TRANS NATIONAL TIMES SACCO', 'UNAITAS SACCO',
       'UNITED NATIONS SACCO', 'VIKTAS SACCO', 'VISION AFRICA SACCO',
       'VISION POINT SACCO', 'WAKULIMA COMMERCIAL SACCO',
       'WANAANGA SACCO',

In [26]:
#confirming the number of saccos captured, we had 52 files
df_all['SACCO'].nunique()

52

In [27]:
#checking if sacco name is present
df_all.head()

,Account_Name,2019,2020,2021,2022,SACCO,2012,2013,2014,2015,2016,2017,2018,2023,2011,2007,2008,2009,2010
0,Total Assets,269371213.0,2.934117e+08,369437115.0,403387182.7,ACUMEN SACCO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Total Deposit liabilities,221613427.0,2.577053e+08,296337292.0,320546208.0,ACUMEN SACCO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Total Equity,33585269.0,3.528895e+07,53706746.0,63405036.0,ACUMEN SACCO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Total Liabilities,235786131.0,2.581184e+08,315730368.0,339982145.0,ACUMEN SACCO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Cash & Cash Equivalent,10834463.0,6.094258e+07,56687339.0,77254741.0,ACUMEN SACCO,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [28]:
#checking the columns
df_all.columns

Index(['Account_Name', '2019', '2020', '2021', '2022', 'SACCO', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2023', '2011', '2007', '2008',
       '2009', '2010'],
      dtype='object')

In [29]:
#removing the junk columns
df_all = df_all.loc[:, ~df_all.columns.str.contains('Unnamed')]

In [30]:
####Column Selection and Data Cleaning

#In this step, we define a list of relevant columns (`valid_cols`) and filter the dataset to retain only those columns.

valid_cols = ['Account_Name', 'SACCO'] + [...]
valid_cols = ['Account_Name', 'SACCO'] + [
    col for col in df_all.columns if col.isdigit()
]

df_all = df_all[valid_cols]

In [31]:
#confirming the columns
df_all.columns

Index(['Account_Name', 'SACCO', '2019', '2020', '2021', '2022', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2023', '2011', '2007', '2008',
       '2009', '2010'],
      dtype='object')

### Reshaping Data to Long Format

The dataset is currently in a wide format, where each year is represented as a separate column.

To prepare the data for structured analysis, it is reshaped into a long format. This transformation:
- Converts year columns into a single variable
- Retains SACCO identity for each observation
- Creates a consistent structure for time-based analysis

The resulting dataset includes:
- SACCO
- Account_Name
- Year
- Value

In [32]:
df_long = df_all.melt(
    id_vars=['Account_Name', 'SACCO'],
    var_name='Year',
    value_name='Value'
)

In [33]:
df_long[['SACCO']].head()

,SACCO
0,ACUMEN SACCO
1,ACUMEN SACCO
2,ACUMEN SACCO
3,ACUMEN SACCO
4,ACUMEN SACCO


### Creating the Final Modeling Dataset

The dataset is currently in long format, where each row represents a financial value for a specific SACCO, account, and year.

To prepare the data for modeling, it is reshaped into a wide format where:
- Each row represents a SACCO at a specific year
- Each column represents a financial feature
- Values correspond to financial amounts

Both SACCO and Year are used as identifiers to ensure that data from different SACCOs and time periods is not incorrectly aggregated.

In [34]:
df_model = df_long.pivot_table(
    index=['SACCO', 'Year'],
    columns='Account_Name',
    values='Value',
    aggfunc='sum'
)

df_model = df_model.reset_index()

df_model.head()

Account_Name,SACCO,Year,Cash & Cash Equivalent,Current Year's Surplus,Financial Investments,Gross Loan Portfolio,Net Loan Portfolio,Share Capital,Statutory Reserve,Total Assets,Total Deposit liabilities,Total Equity,Total Liabilities
0,ACUMEN SACCO,2007,0,0,0,0,0,0,0,0,0,0,0
1,ACUMEN SACCO,2008,0,0,0,0,0,0,0,0,0,0,0
2,ACUMEN SACCO,2009,0,0,0,0,0,0,0,0,0,0,0
3,ACUMEN SACCO,2010,0,0,0,0,0,0,0,0,0,0,0
4,ACUMEN SACCO,2011,0,0,0,0,0,0,0,0,0,0,0


In [35]:
df_model.shape
df_model.columns[:50]

Index(['SACCO', 'Year', 'Cash & Cash Equivalent', 'Current Year's Surplus',
       'Financial Investments', 'Gross Loan Portfolio', 'Net Loan Portfolio',
       'Share Capital', 'Statutory Reserve', 'Total Assets',
       'Total Deposit liabilities', 'Total Equity', 'Total Liabilities'],
      dtype='object', name='Account_Name')

Dataset Structure Inspection

In this step, we explore the structure of the modeling dataset (`df_model`) by checking its dimensions and previewing the column names.
We want to ensure no repetition.

Feature Selection

After preparing the dataset, relevant financial variables are selected for modeling.

The target variable is:
- Current Year's Surplus

This represents SACCO performance and is used as a proxy for financial returns.

The selected features include key financial indicators such as:
- Total Assets
- Loan Portfolio
- Deposits
- Equity
- Investments

These variables are expected to influence SACCO performance and are used as predictors in the model.

In [36]:
#Defining the target
y = df_model["Current Year's Surplus"]

In [37]:
#Defining the features
features = [
    'Total Assets',
    'Total Deposit liabilities',
    'Total Equity',
    'Total Liabilities',
    'Cash & Cash Equivalent',
    'Financial Investments',
    'Gross Loan Portfolio',
    'Net Loan Portfolio',
    'Share Capital',
    'Statutory Reserve'
]

X = df_model[features]

In [38]:
#checking missing values
X.isna().sum()

Account_Name
Total Assets                 0
Total Deposit liabilities    0
Total Equity                 0
Total Liabilities            0
Cash & Cash Equivalent       0
Financial Investments        0
Gross Loan Portfolio         0
Net Loan Portfolio           0
Share Capital                0
Statutory Reserve            0
dtype: int64

Baseline Model

A baseline regression model is built to predict SACCO performance (Current Year's Surplus) using selected financial indicators.

This provides a starting point for evaluating model performance and understanding relationships between variables.

## Baseline Model: Linear Regression

A Linear Regression model was used as the baseline model for this analysis.

Baseline models provide a simple and interpretable starting point for evaluating model performance.

Linear Regression was selected because:
- It is easy to implement and interpret
- It provides a benchmark for comparison with more complex models
- It helps identify basic relationships between financial variables and SACCO performance

The performance of this model is later compared with more advanced machine learning models.

### Splitting the Dataset

The dataset is divided into training and testing sets to evaluate model performance.

- Training set: used to train the model
- Testing set: used to evaluate model performance on unseen data

In [39]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

In [40]:
X_train.shape, X_test.shape

((707, 10), (177, 10))

### Training the Model

A Linear Regression model is trained using the training dataset.

The model learns relationships between financial indicators and SACCO performance (Current Year's Surplus).

In [41]:
#confirming non numeric columns first
for col in X.columns:
    if X[col].dtype == 'object':
        print(col)

Total Assets
Total Deposit liabilities
Total Equity
Total Liabilities
Cash & Cash Equivalent
Financial Investments
Gross Loan Portfolio
Net Loan Portfolio
Share Capital
Statutory Reserve


In [42]:
#replacing non-numeric columns on features
X = X.apply(pd.to_numeric, errors='coerce')

In [43]:
#replacing non-numeric columns on the target
y = pd.to_numeric(y, errors='coerce')

In [44]:
#confirming missing values
X.isna().sum()

Account_Name
Total Assets                 1
Total Deposit liabilities    0
Total Equity                 0
Total Liabilities            3
Cash & Cash Equivalent       0
Financial Investments        0
Gross Loan Portfolio         0
Net Loan Portfolio           0
Share Capital                1
Statutory Reserve            0
dtype: int64

### Rebuilding Baseline Model with Improved Missing Value Handling

The baseline model is reconstructed using median imputation instead of zero filling.

Median is used because financial data is often skewed and contains outliers, making it a more robust measure for handling missing values.

In [45]:
#fill missing values with median
X = X.fillna(X.median())
y = y.fillna(y.median())

In [46]:
#resplitting the data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

In [47]:
#training the model
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

### Making Predictions

The trained model is used to predict SACCO performance on the test dataset.

These predictions are then compared with actual values to evaluate model performance.

In [48]:
y_pred = model.predict(X_test)

In [49]:
y_pred[:5]

array([-7.27425118e+06,  1.68590786e+07, -3.43282572e+06, -6.70204703e+06,
        1.21430105e+08])

### Model Evaluation

The performance of the model is evaluated using standard regression metrics:

- Mean Absolute Error (MAE): average prediction error
- Mean Squared Error (MSE): penalizes larger errors
- R-squared (R²): proportion of variance explained by the model

In [50]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mae, mse, r2

(24139051.228664305, 5746149335671352.0, 0.47955740925104373)

### Feature Scaling

Financial variables have different magnitudes (e.g., assets vs cash balances).

To improve model performance, features are standardized to ensure they are on a similar scale.

In [51]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

In [52]:
#resplit
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y,
    test_size=0.2,
    random_state=42
)

In [53]:
#retrain
model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

Baseline Model Evaluation (Before Transformation & Imputation)

In this step, we evaluate the performance of the initial model using standard regression metrics:

```python
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mae, mse, r2

In [54]:
#reevaluate
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mae, mse, r2

(24139051.22866449, 5746149335671310.0, 0.4795574092510475)

Baseline Model Performance & Next Steps

We evaluated the initial model using key regression metrics:

- **MAE:** 24,139,051.23  
- **MSE:** 5.75 × 10¹⁵  
- **R²:** 0.48  

---

Interpretation of Results

The model shows **moderate explanatory power** but poor predictive accuracy:

- The **R² score (0.48)** indicates the model explains about 48% of the variance  
  → This is **not terrible**, but leaves significant room for improvement  

- The **MAE is very high**, meaning predictions are off by large margins on average  
  → Errors are large in absolute financial terms  

- The **MSE is extremely large**, which highlights:
  - Presence of **outliers**
  - Large variance in the target variable  
  - Sensitivity to extreme values (since MSE penalizes large errors more)

---

Key Issues Identified

1. **Skewed Target Distribution**
   - Financial variables (e.g., assets, loans) are typically **right-skewed**
   - This causes instability in model predictions

2. **Outliers**
   - Large institutions likely dominate the scale
   - Model struggles to generalize across different magnitudes

3. **Poor Missing Value Strategy**
   - Filling with zeros can distort relationships
   - Especially problematic in financial datasets

4. **Scale Imbalance**
   - Features may exist on vastly different scales
   - Impacts model learning and error metrics

---

Recommended Next Steps

To improve model performance and stability:

1. Apply Log Transformation
- Transform the target variable:
  ```python
  y = np.log1p(y)

### Log-Transformed Model with Median Imputation

To improve model performance, a log transformation is applied to reduce skewness in financial data.

Missing values are handled using median imputation to ensure robustness against outliers.

## Improved Model: Log-Transformed Linear Regression

To improve the baseline model, data preprocessing techniques were applied:

- Log transformation was used to reduce skewness in financial variables
- Median imputation was applied to handle missing values robustly

These improvements aim to stabilize variance and improve model performance.

In [69]:
#ensure numbers are numeric
X = X.apply(pd.to_numeric, errors='coerce')
y = pd.to_numeric(y, errors='coerce')

In [70]:
import numpy as np

X_log = np.log1p(X)
y_log = np.log1p(y)

c:\Users\morin\anaconda3\Lib\site-packages\pandas\core\internals\blocks.py:351: RuntimeWarning: invalid value encountered in log1p
  result = func(self.values, **kwargs)
c:\Users\morin\anaconda3\Lib\site-packages\pandas\core\arraylike.py:402: RuntimeWarning: invalid value encountered in log1p
  result = getattr(ufunc, method)(*inputs, **kwargs)


In [71]:
#check missing values
X_log.isna().sum()

Account_Name
Total Assets                  0
Total Deposit liabilities     0
Total Equity                 10
Total Liabilities             0
Cash & Cash Equivalent        1
Financial Investments         0
Gross Loan Portfolio          0
Net Loan Portfolio            0
Share Capital                 0
Statutory Reserve             0
dtype: int64

In [72]:
#fill in missing values with median
X_log = X_log.fillna(X_log.median())
y_log = y_log.fillna(y_log.median())

In [73]:
#check missing values again, should be none.
X_log.isna().sum()

Account_Name
Total Assets                 0
Total Deposit liabilities    0
Total Equity                 0
Total Liabilities            0
Cash & Cash Equivalent       0
Financial Investments        0
Gross Loan Portfolio         0
Net Loan Portfolio           0
Share Capital                0
Statutory Reserve            0
dtype: int64

In [74]:
#resplit
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_log, y_log,
    test_size=0.2,
    random_state=42
)

In [75]:
#retrain
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [76]:
### 🔮 Model Prediction (After Log Transformation)

#After applying log transformation and preprocessing the data, we generate predictions on the test set:

y_pred = model.predict(X_test)

In [77]:
#evaluate the model
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mae, mse, r2

(1.3341224927757245, 9.986820393980885, 0.7920548298955852)

### Baseline Model Performance

After applying log transformation and median imputation, the model showed significant improvement.

- R² improved from a negative value to approximately 0.79
- Prediction errors reduced substantially

This demonstrates the importance of handling skewed financial data and outliers appropriately.

The model is considered reliable for analyzing SACCO financial performance.

In [64]:
coefficients = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
})

coefficients.sort_values(by='Coefficient', ascending=False)

,Feature,Coefficient
6,Gross Loan Portfolio,0.872655
5,Financial Investments,0.609495
4,Cash & Cash Equivalent,0.267113
2,Total Equity,0.247414
9,Statutory Reserve,0.209594
8,Share Capital,0.139584
3,Total Liabilities,0.105630
0,Total Assets,0.029706
7,Net Loan Portfolio,-0.050264
1,Total Deposit liabilities,-1.528736


## Key Insights

The model identified the most important drivers of SACCO financial performance:

- Gross Loan Portfolio is the strongest predictor of surplus, highlighting the importance of lending activities.
- Financial Investments also contribute significantly, indicating the value of diversified income sources.
- Statutory Reserves show a positive relationship with performance, suggesting that financial stability supports growth.

Interestingly, Total Deposit Liabilities showed a negative relationship with surplus, implying that high deposits may reduce profitability if not effectively utilized for income-generating activities.

These findings suggest that SACCOs should focus on efficient loan management and investment strategies to improve performance.

Baseline Model- A simple, interpretable model used as a reference point

### Ridge Regression Model

Ridge Regression is applied as an advanced model to improve the stability of predictions.

It introduces L2 regularization, which penalizes large coefficients and helps reduce overfitting.

This is particularly useful for financial datasets where variables may be highly correlated.

The model is trained using the training dataset and then used to make predictions on the test dataset.

In [ ]:
#import Ridge
from sklearn.linear_model import Ridge

ridge = Ridge(alpha=1.0)

ridge.fit(X_train, y_train)

y_pred_ridge = ridge.predict(X_test)

In [66]:
#evaluate the model
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae_ridge = mean_absolute_error(y_test, y_pred_ridge)
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
r2_ridge = r2_score(y_test, y_pred_ridge)

mae_ridge, mse_ridge, r2_ridge

(1.334615756678289, 9.99778541230635, 0.7918265167277364)

### Ridge Regression Performance

Ridge Regression achieved strong performance, with an R² of approximately 0.79.

However, it did not outperform the improved Linear Regression model.

This suggests that regularization was not necessary, as the log transformation had already addressed issues related to skewness and instability.

The Linear Regression model remains the better-performing model in this case.

### Lasso Regression

Lasso Regression is applied to perform feature selection by shrinking less important feature coefficients to zero.

This helps identify the most important predictors of SACCO performance.

In [67]:
#import Lasso
from sklearn.linear_model import Lasso

lasso = Lasso(alpha=0.01)

lasso.fit(X_train, y_train)

y_pred_lasso = lasso.predict(X_test)

c:\Users\morin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.770e+03, tolerance: 3.614e+00
  model = cd_fast.enet_coordinate_descent(


In [68]:
#evaluating the model
mae_lasso = mean_absolute_error(y_test, y_pred_lasso)
mse_lasso = mean_squared_error(y_test, y_pred_lasso)
r2_lasso = r2_score(y_test, y_pred_lasso)

mae_lasso, mse_lasso, r2_lasso

(1.3443371093183472, 10.406300195299622, 0.7833204384477161)

## Advanced Model: Random Forest Regression

To further enhance the analysis, a Random Forest Regression model was applied.

Random Forest is an ensemble learning method that builds multiple decision trees and combines their predictions to improve accuracy and robustness.

This model is particularly suitable for financial data because:
- It captures non-linear relationships between variables
- It is less sensitive to outliers
- It reduces overfitting through averaging multiple trees

The Random Forest model is trained using the training dataset and evaluated on the test dataset to compare its performance with linear models.

In [79]:
#importing Random forest
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

In [80]:
#evaluating the model
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

mae_rf, mse_rf, r2_rf

(1.0610814168474583, 8.599428885058927, 0.8209430397504535)

## Final Model Selection

Several models were evaluated:

- Linear Regression (baseline and improved)
- Ridge Regression
- Lasso Regression
- Random Forest Regression

The Random Forest model achieved the best performance:

- Highest R² (~0.82)
- Lowest prediction error

This indicates that SACCO financial performance is influenced by non-linear relationships between variables.

Therefore, Random Forest was selected as the final model.

The superior performance of Random Forest suggests that SACCO financial outcomes are not purely linear and involve complex interactions between financial indicators.

In [81]:
importance = pd.DataFrame({
    'Feature': X.columns,
    'Importance': rf.feature_importances_
}).sort_values(by='Importance', ascending=False)

importance

,Feature,Importance
3,Total Liabilities,0.310176
2,Total Equity,0.282533
5,Financial Investments,0.187137
4,Cash & Cash Equivalent,0.069009
9,Statutory Reserve,0.053315
6,Gross Loan Portfolio,0.031431
1,Total Deposit liabilities,0.028146
8,Share Capital,0.017558
0,Total Assets,0.012480
7,Net Loan Portfolio,0.008215


## Feature Importance Analysis (Random Forest)

The Random Forest model identified the most important factors influencing SACCO performance:

- Total Liabilities emerged as the most significant predictor
- Total Equity also showed strong influence
- Financial Investments contributed meaningfully to performance

These findings suggest that SACCO financial performance is strongly influenced by balance sheet structure and capital management.

Compared to linear models, Random Forest highlights more complex interactions between financial variables.

The results indicate that SACCO performance is driven not only by income-generating activities such as lending, but also by overall financial structure and capital management.

## Reproducible Model Deployment

To ensure reproducibility, the deployment process follows the exact same preprocessing steps used during model training.

This includes:
- Maintaining consistent feature ordering
- Applying identical log transformations
- Using the trained Random Forest model

This guarantees consistent and reliable predictions for new input data.

In [82]:
feature_order = [
    'Total Assets',
    'Total Deposit liabilities',
    'Total Equity',
    'Total Liabilities',
    'Cash & Cash Equivalent',
    'Financial Investments',
    'Gross Loan Portfolio',
    'Net Loan Portfolio',
    'Share Capital',
    'Statutory Reserve'
]

In [83]:
def predict_surplus(input_data):
    import pandas as pd
    import numpy as np
    
    # create DataFrame with correct order
    input_df = pd.DataFrame([input_data])[feature_order]
    
    # apply SAME transformation
    input_log = np.log1p(input_df)
    
    # predict
    prediction_log = rf.predict(input_log)
    
    # reverse log
    prediction = np.expm1(prediction_log)
    
    return prediction[0]

In [84]:
import joblib

joblib.dump(rf, "random_forest_model.pkl")

['random_forest_model.pkl']

In [85]:
rf = joblib.load("random_forest_model.pkl")

### Deployment Reproducibility

The deployment pipeline ensures reproducibility by maintaining consistent preprocessing steps and feature structure.

This guarantees that predictions remain stable and reliable across different runs and environments.

In [88]:
sample_input = {
    'Total Assets': 800000000,
    'Total Deposit liabilities': 500000000,
    'Total Equity': 150000000,
    'Total Liabilities': 650000000,
    'Cash & Cash Equivalent': 80000000,
    'Financial Investments': 120000000,
    'Gross Loan Portfolio': 400000000,
    'Net Loan Portfolio': 350000000,
    'Share Capital': 100000000,
    'Statutory Reserve': 50000000
}

In [87]:
predict_surplus(sample_input)

3268636.5411408953

### Example Prediction

A sample SACCO financial profile was used to test the deployed model.

The model successfully generated a predicted surplus value, demonstrating its practical applicability in financial forecasting.

## Deployment Results

A sample SACCO financial profile was used to test the deployed model.

The model predicted a surplus of approximately 3.27 million.

This demonstrates that the model can be used to estimate financial performance based on key financial indicators.

Such a system can support SACCO managers in forecasting and decision-making.

In [89]:
small_sacco = {
    'Total Assets': 200000000,
    'Total Deposit liabilities': 120000000,
    'Total Equity': 40000000,
    'Total Liabilities': 160000000,
    'Cash & Cash Equivalent': 20000000,
    'Financial Investments': 30000000,
    'Gross Loan Portfolio': 100000000,
    'Net Loan Portfolio': 90000000,
    'Share Capital': 25000000,
    'Statutory Reserve': 15000000
}

predict_surplus(small_sacco)

16500.015673365266

### Additional Deployment Testing

The model was tested using a smaller SACCO profile.

The predicted surplus was significantly lower (~16,500) compared to a larger SACCO (~3.27 million).

This demonstrates that the model appropriately scales predictions based on SACCO size and financial structure.

It also highlights the sensitivity of financial performance to key variables such as assets, loans, and investments.